src/drivers/intel/cavs/hda-dma.c

```diff
@@ -387,25 +398,26 @@ static int hda_dma_host_copy(struct dma *dma, int channel, int bytes,
 
 	hda_dma_get_dbg_vals(chan, HDA_DBG_PRE, HDA_DBG_HOST);
 
-	if (flags & DMA_COPY_PRELOAD) {
+	if (!(flags & DMA_COPY_PRELOAD))
+		/* set BFPI to let host gateway know we have read size,
+		 * which will trigger next copy start.
+		 */
+		hda_dma_inc_fp(dma, chan->index, bytes);
+
+	/* blocking mode copy */
+	if (flags & DMA_COPY_BLOCKING) {
 		deadline = platform_timer_get(platform_timer) +
 			clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) *
 			PLATFORM_HOST_DMA_TIMEOUT / 1000;
 
-		/* waiting for buffer full */
-		while (!(host_dma_reg_read(dma, chan->index, DGCS) &
```

Comment:
So this is not going to check status registers?

Reply:
It will. See implementation of hda_dma_avail_data_size and hda_dma_free_data_size.